### DISCUSSION TITLE
What are the technical guidelines in writing data-protocols & extensions?

### DISCUSSION THREAD

**Discussion by oxisto** [eclipse-edc/.github#41] [Q&A] [answered]
Hi Eclipse-DataSpaceCommunity!

We are interested in bringing over some of the functionality from our Trusted Connector and IDSCP2 projects (https://github.com/industrial-data-space/trusted-connector, https://github.com/industrial-data-space/idscp2-java) onto this platform. We are also interested how we can contribute to some of the aspects of SSI (see #56).

Therefore, we would like to start the discussion about some of the technical guidelines when writing data-protocols and extensions.

1. I understand that you want to minimise the number of dependencies in the core. Does that also apply to the extensions? I do see some extensions with quite a heavy load of dependencies, e.g. to the different Cloud providers
2. Are you currently set on *Java* as the *only* language for extensions? Since recently we basically ported most of our code from Java to Kotlin (while still running on the JVM). So porting it back to Java is not really an option. A lot of our modules are already available as individual dependency libraries, but allowing the use of Kotlin in the extension itself would probably give us a little bit more freedom.

We wanted to clarify these basic points before coming up with a plan to contribute (and what to contribute) in more detail.

cc @gbrost @schanzen

**Comment by jimmarino** [comment, marked as answer]
Hi

We would welcome contributions in any area.

Regarding your questions:

1. Our idea is to restrict core dependencies and commonly used extensions but allow more leeway in other extensions. Overall, however, we would like to minimize dependencies for modules in this repo since the build dependency tree has to remain manageable as the number of extensions grows. Unfortunately, some cloud providers (e.g. the AWS and MSAL libraries) contain a boatload of dependencies. I don't think we can do much with the AWS libraries, but we should see if we can eliminate the use of MSAL.

2. For Kotlin, I would start by saying that the existing code from the Trusted Connector will likely have to be significantly refactored since the EDC architecture is entirely different from previous connector implementations.  Also, I think it is important that the codebase is consistent and maintainable by the community on a long-term basis. We have settled on Java as the runtime language since that has a large open-source developer community. My preference would be to continue with that approach and not introduce barriers (e.g. knowledge of Kotlin) for people to contribute and maintain areas of our codebase.

I think porting the Trusted Connector to EDC modules would be great, and doing it in Java would be a way to make sure those extensions are integrated with the rest of the EDC community.

**Reply by oxisto** [reply]
> Hi
> 
> We would welcome contributions in any area.
> 
> Regarding your questions:
> 
> 1. Our idea is to restrict core dependencies and commonly used extensions but allow more leeway in other extensions. Overall, however, we would like to minimize dependencies for modules in this repo since the build dependency tree has to remain manageable as the number of extensions grows. Unfortunately, some cloud providers (e.g. the AWS and MSAL libraries) contain a boatload of dependencies. I don't think we can do much with the AWS libraries, but we should see if we can eliminate the use of MSAL.
> 2. For Kotlin, I would start by saying that the existing code from the Trusted Connector will likely have to be significantly refactored since the EDC architecture is entirely different from previous connector implementations.  Also, I think it is important that the codebase is consistent and maintainable by the community on a long-term basis. We have settled on Java as the runtime language since that has a large open-source developer community. My preference would be to continue with that approach and not introduce barriers (e.g. knowledge of Kotlin) for people to contribute and maintain areas of our codebase.
> 
> I think porting the Trusted Connector to EDC modules would be great, and doing it in Java would be a way to make sure those extensions are integrated with the rest of the EDC community.

Thanks for your answers. Just to avoid any misunderstandings, Kotlin still uses the JVM as a *runtime*. It is just basically a different "frontend" language. Having Kotlin and Java side-by-side in the same project is pretty straight forward. We had been using it during our migration to Kotlin.

Re-creating huge parts of the Kotlin implementation back into Java is probably not very efficient for us. So the way I would approach it is probably to package existing parts of our connector implementation into Maven libraries (still based on Kotlin, but external to this project) and use them here using an extension, which probably has a small Java stub. We will discuss this internally and come back to you on this.

**Reply by jimmarino** [reply]
Thanks, I'm familiar with Kotlin. 

One option you could have is to host your extensions in your existing project and have them rely on EDC modules via Maven Central artifacts. We need to start publishing there soon, so this would be a good use case.

I suspect you will still need to refactor your codebase to have them function as extensions in EDC.

We're happy to help, so let us know

**Comment by oxisto** [comment]
Maybe another thought on this. @jimmarino Would it be possible / is it planned to publish jar artifacts of the EDSC on GH or maven central, so others can use it to base their connector implementation on. 

Currently, we use SpringBoot in the Trusted Connector, but this was mostly because of the fact the the IDSA-DSC was using it and we wanted to be more compatible. If now things move towards the EDSC, we could think about switching the underlying "framework" from SpringBoot to the EDSC framework, basically keeping our code in an separate repo and creating our extensions in Kotlin (at least for now).

Just thinking out loud here.

cc @gbrost

**Reply by jimmarino** [reply]
Yeah, that word work. Can you create a PR for that and we'll look into it?
